cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp

```diff
@@ -89,6 +89,82 @@ std::vector<NT::fr> calculate_contract_leaves(BaseRollupInputs baseRollupInputs)
     return contract_leaves;
 }
 
+template <size_t N>
+NT::fr iterate_through_tree_via_sibling_path(NT::fr leaf,
+                                             NT::uint32 leafIndex,
+                                             std::array<NT::fr, N> siblingPath,
+                                             size_t startAtDepth = 0)
+{
+    for (size_t i = startAtDepth; i < siblingPath.size(); i++) {
+        if (leafIndex & (1 << i)) {
+            leaf = crypto::pedersen_hash::hash_multiple({ leaf, siblingPath[i] });
+        } else {
+            leaf = crypto::pedersen_hash::hash_multiple({ siblingPath[i], leaf });
+        }
+    }
+    return leaf;
+}
+
+template <size_t N>
+void check_membership_of_subtree_in_snapshot(NT::uint32 depth,
+                                             NT::uint32 nextAvailableLeafIndex,
+                                             std::array<NT::fr, N> siblingPath,
+                                             NT::fr snapshotRoot)
+{
+    stdlib::merkle_tree::MemoryTree empty_subtree = stdlib::merkle_tree::MemoryTree(depth);
```

Comment:
We should compute the emptry tree roots separately and just have them as "constants", mainly to not have to rebuild a lot of unneeded trees. It will probably be fine to have in a test or something like that, add a comment that we should move it at to that.